### PR TITLE
Support of JsonPolymorphic and JsonDerivedType attributes

### PR DESCRIPTION
**Pull request by k0ka**
## The issue or feature being addressed

supersedes #2671 

## Details on the issue fix or feature implementation

- Added checks for `System.Text.Json` polymorphic attributes in the same handlers, where swagger native attributes are handled. Swagger attributes have priority over `System.Text.Json` ones.

- Added this case to `NSwagClientExample` integration test

- Refactored the way `Basic` and `NSwagClientExample` tests run. They both have different outputs for .net6 and .net8/9. I made a different theory for such tests in .net6 mode. The redudant `SwaggerEndpoint_ReturnsValidSwaggerJson_Basic_DotNet_8` was deleted, as it is handled by the main `SwaggerEndpoint_ReturnsValidSwaggerJson`

### CODE REVIEW

File: src/Swashbuckle.AspNetCore.Annotations/AnnotationsSwaggerGenOptionsExtensions.cs [hunk 1]
```diff
@@ -75,6 +76,16 @@ private static IEnumerable<Type> AnnotationsSubTypesSelector(Type type)
                 return obsoleteAttribute.SubTypes;
             }
 
+#if NET7_0_OR_GREATER
+            var jsonDerivedTypeAttributes = type.GetCustomAttributes(false)
+                .OfType<JsonDerivedTypeAttribute>().ToList();
+
+            if (jsonDerivedTypeAttributes.Any())
```
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.Annotations/AnnotationsSwaggerGenOptionsExtensions.cs:83]:
```suggestion
                .OfType<JsonDerivedTypeAttribute>()
                .ToList();

            if (jsonDerivedTypeAttributes.Count > 0)
```

File: src/Swashbuckle.AspNetCore.Annotations/AnnotationsSwaggerGenOptionsExtensions.cs [hunk 2]
```diff
@@ -100,6 +111,16 @@ private static string AnnotationsDiscriminatorNameSelector(Type baseType)
                 return obsoleteAttribute.Discriminator;
             }
 
+#if NET7_0_OR_GREATER
+            var jsonPolymorphicAttributes = baseType.GetCustomAttributes(false)
+                .OfType<JsonPolymorphicAttribute>().FirstOrDefault();
```
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.Annotations/AnnotationsSwaggerGenOptionsExtensions.cs:116]:
```suggestion
                .OfType<JsonPolymorphicAttribute>()
                .FirstOrDefault();
```

File: test/Swashbuckle.AspNetCore.IntegrationTests/SwaggerVerifyIntegrationTest.cs [hunk 3]
```diff
@@ -40,30 +40,23 @@ public async Task SwaggerEndpoint_ReturnsValidSwaggerJson(
...
+            Type startupType,
+            string swaggerRequestUri)
         {
             var testSite = new TestSite(typeof(Basic.Startup));
             using var client = testSite.BuildClient();
 
             using var swaggerResponse = await client.GetAsync("/swagger/v1/swagger.json");
             var swagger = await swaggerResponse.Content.ReadAsStringAsync();
-            await Verifier.VerifyJson(swagger);
+            await Verifier.Verify(swagger).UseParameters(startupType, GetVersion(swaggerRequestUri));
```
Comment (martincostello) [inline test/Swashbuckle.AspNetCore.IntegrationTests/SwaggerVerifyIntegrationTest.cs:55]: Why change this from `VerifyJson`?

File: test/WebSites/NswagClientExample/Controllers/SystemTextJsonAnimalsController.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+using Microsoft.AspNetCore.Mvc;
+
+namespace NSwagClientExample.Controllers
+{
+#if NET7_0_OR_GREATER
+    [ApiController]
+    [Route("[controller]")]
+    public class SystemTextJsonAnimalsController : ControllerBase
+    {
+        [HttpPost]
```
Comment (martincostello) [inline test/WebSites/NswagClientExample/Controllers/SystemTextJsonAnimalsController.cs:13]: Could we add `[Produces(...)]` here for JSON? Then we shouldn't(?) get three different content-types that are all the basically the same in the responses in the snapshot.

File: test/WebSites/NswagClientExample/swagger_net8.0.json [hunk 5]
```diff
@@ -134,6 +134,38 @@
...
+            },
+            "text/json": {
+              "schema": {
+                "$ref": "#/components/schemas/SystemTextJsonAnimal"
+              }
+            },
+            "application/*+json": {
+              "schema": {
+                "$ref": "#/components/schemas/SystemTextJsonAnimal"
+              }
```
Comment (martincostello) [inline test/WebSites/NswagClientExample/swagger_net8.0.json:158]: Like here - we have three types of JSON.

File: test/Swashbuckle.AspNetCore.IntegrationTests/SwaggerVerifyIntegrationTest.cs (hunk 3 above)
Comment (k0ka) [reply test/Swashbuckle.AspNetCore.IntegrationTests/SwaggerVerifyIntegrationTest.cs:55]: the `SwaggerEndpoint_ReturnsValidSwaggerJson` uses `Verify`. I guess there is not much difference between them and `Verify` is more strict as it compares byte-by-byte. We might change to `VerifyJson` there, but I guess it's better to change it to `VerifyJson` in the main test as well (`SwaggerEndpoint_ReturnsValidSwaggerJson`) for consistency.

File: test/WebSites/NswagClientExample/Controllers/SystemTextJsonAnimalsController.cs (hunk 4 above)
Comment (k0ka) [reply test/WebSites/NswagClientExample/Controllers/SystemTextJsonAnimalsController.cs:13]: added it.

File: test/Swashbuckle.AspNetCore.IntegrationTests/SwaggerVerifyIntegrationTest.cs (hunk 3 above)
Comment (martincostello) [reply test/Swashbuckle.AspNetCore.IntegrationTests/SwaggerVerifyIntegrationTest.cs:55]: For consistency and best experience with Verify we should consistently use `VerifyJson` - using just `Verify` was probably an oversight.

Comment (martincostello) [reply test/Swashbuckle.AspNetCore.IntegrationTests/SwaggerVerifyIntegrationTest.cs:55]: Ah OK, I didn't realise that.

**Review by martincostello — APPROVED** [review]

**Comment by martincostello** [comment]
Before I merge this, could you copy-over the changes to the README about how this works please?

https://github.com/domaindrivendev/Swashbuckle.AspNetCore/pull/2671/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5

**Comment by k0ka** [comment]
added Readme

File: README.md [hunk 6]
```diff
@@ -1166,7 +1166,7 @@ services.AddSwaggerGen(c =>
 });
 ```
 
-_NOTE: If you're using the [Swashbuckle Annotations library](#swashbuckleaspnetcoreannotations), it contains a custom selector that's based on the presence of `SwaggerSubType` attributes on base class definitions. This way, you can use simple attributes to explicitly list the inheritance and/or polymorphism relationships you want to expose. To enable this behavior, check out the [Annotations docs](#list-known-subtypes-for-inheritance-and-polymorphism)._
+_NOTE: If you're using the [Swashbuckle Annotations library](#swashbuckleaspnetcoreannotations), it contains a custom selector that's based on the presence of `JsonDerivedType` (or `SwaggerSubType` for .NET 6 or earlier) attributes on base class definitions. This way, you can use simple attributes to explicitly list the inheritance and/or polymorphism relationships you want to expose. To enable this behavior, check out the [Annotations docs](#list-known-subtypes-for-inheritance-and-polymorphism)._
```
Comment (martincostello) [inline README.md:1169]:
```suggestion
> [!NOTE]
> If you're using the [Swashbuckle Annotations library](#swashbuckleaspnetcoreannotations), it contains a custom selector that's based on the presence of `[JsonDerivedType]` (or `[SwaggerSubType]` for .NET 6 or earlier) attributes on base class definitions. This way, you can use simple attributes to explicitly list the inheritance and/or polymorphism relationships you want to expose. To enable this behavior, check out the [Annotations docs](#list-known-subtypes-for-inheritance-and-polymorphism).
```

File: README.md [hunk 7]
```diff
@@ -1232,7 +1232,7 @@ services.AddSwaggerGen(c =>
 });
 ```
 
-_NOTE: If you're using the [Swashbuckle Annotations library](#swashbuckleaspnetcoreannotations), it contains custom selector functions that are based on the presence of `SwaggerDiscriminator` and `SwaggerSubType` attributes on base class definitions. This way, you can use simple attributes to explicitly provide discriminator metadata. To enable this behavior, check out the [Annotations docs](#enrich-polymorphic-base-classes-with-discriminator-metadata)._
+_NOTE: If you're using the [Swashbuckle Annotations library](#swashbuckleaspnetcoreannotations), it contains custom selector functions that are based on the presence of `JsonPolymorphic` (or `SwaggerDiscriminator` for .NET 6 or earlier) and `JsonDerivedType` (or `SwaggerSubType` for .NET 6 or earlier) attributes on base class definitions. This way, you can use simple attributes to explicitly provide discriminator metadata. To enable this behavior, check out the [Annotations docs](#enrich-polymorphic-base-classes-with-discriminator-metadata)._
```
Comment (martincostello) [inline README.md:1235]:
```suggestion
> [!NOTE]
> If you're using the [Swashbuckle Annotations library](#swashbuckleaspnetcoreannotations), it contains custom selector functions that are based on the presence of `[JsonPolymorphic]` (or `[SwaggerDiscriminator]` for .NET 6 or earlier) and `[JsonDerivedType]` (or `[SwaggerSubType]` for .NET 6 or earlier) attributes on base class definitions. This way, you can use simple attributes to explicitly provide discriminator metadata. To enable this behavior, check out the [Annotations docs](#enrich-polymorphic-base-classes-with-discriminator-metadata).
```

**Comment by martincostello** [comment]
Resolves #2571.

**Comment by alainkaiser** [comment]
Yeah, this is actually nice! Thanks guys!

**Comment by schnerring** [comment]
Thanks for picking up my PR and finishing the work. Good job!

**Pull request merged**